_posts/2018-02-07-active-record-5-1-api-changes.markdown

```diff
@@ -0,0 +1,149 @@
+changes between these two versions. These changes are best explained in [this commit](https://github.com/rails/rails/commit/16ae3db5a5c6a08383b974ae6c96faac5b4a3c81),
+and the best aspect of it IMHO is explicitness.
+
+It reduces the ambiguity of simply using `changed?`, which would return `true`
+when the object was dirty, not yet saved to the database (within `before_save`),
+and `true` after the object was saved to the database (within `after_save`).
+
+The second case will *no longer* be `true`, but rather `false`. To reduce
+ambiguity now, you can use the more explicit `saved_changes?` method to ask
+whether the object has been changed and if its changes were saved to the database.
+
+## New behavior
+
+A developer's intentions will become more clear with the behavior changes
+introduced in Rails 5.2:
+
+```ruby
+2.4.2 :003 > user.email = "mauro@ombulabs.com"
+ => "mauro@ombulabs.com"
+2.4.2 :004 > user.save
+
+From: /Users/mauro-oto/Projects/rails52test/app/models/user.rb @ line 2 :
+
+    1: class User < ApplicationRecord
+ => 2:   after_save { binding.pry }
+    3: end
+
+[1] pry(#<User>)> email_changed?
+=> false
+[2] pry(#<User>)> changed?
+=> false
+[3] pry(#<User>)> changes
+=> {}
+[4] pry(#<User>)> previous_changes
+=> {"email"=>["john@doe.org", "mauro@ombulabs.com"]}
+[5] pry(#<User>)> saved_change_to_email?
+=> true
+[6] pry(#<User>)> saved_change_to_email
+=> ["john@doe.org", "mauro@ombulabs.com"]
+[7] pry(#<User>)> email_before_last_save
+=> "john@doe.org"
+```
+
+`attribute_changed?` and `changed?` remain the same when called within a
+`before_save` callback. However, the Rails core team has also introduced a set
+of methods which can help further reduce the potential ambiguity of `changed?`
+before saving to the database: `will_save_change_to_attribute?` and
+`has_changes_to_save?`:
+
+```ruby
+2.4.2 :016 > user.email = "mauro@ombulabs.com"
+ => "mauro@ombulabs.com"
+2.4.2 :017 > user.has_changes_to_save?
+ => true
+2.4.2 :018 > user.will_save_change_to_email?
+ => true
+2.4.2 :019 > user.save
+ => true
+2.4.2 :020 > user.will_save_change_to_email?
+ => nil
+2.4.2 :021 > user.has_changes_to_save?
+ => false
+```
+
+In conclusion, if you use these methods and are planning to migrate to Rails 5.2
+or even 5.1, you should take these changes into account:
+
+### Must do:
+
+After modifying an object and after saving to the database, or within `after_save`:
+- `attribute_changed?` should now be `saved_change_to_attribute?`
+- `changed?` should now be `saved_changes?`
+- `changes` should now be `saved_changes`
+- `previous_changes` has no replacement, since the behavior for it changes.
+
+### Optional (less ambiguity, more readable, but longer):
+
+After modifying an object and before saving to the database, or within `before_save`:
+- `attribute_changed?` should now be `will_save_change_to_attribute?`
+- `changed?` should now be `has_changes_to_save?`
+- `changes` should now be `changes_to_save`
```

Comment:
@mauro-oto Maybe you can add a snippet here that mentions our upgrade series of articles and which has a reference to `fastruby.io`